### ISSUE TITLE
idrop is bigger than W in triageTemplates2

### ISSUE THREAD

**Issue by NoahDolev** [MouseLand/Kilosort2#44]
Hi,

I get an error in triageTemplates2. Apparently idrop = nsp<m0 leads to a vector greater than the second dimension of W. 

```{Matlab}
299.90 sec, 1 / 4931 batches, 1008 units, nspks: 16.5796, mu: 10.0000, nst0: 0, merges: 0.1000, 0.0000 
Error using gpuArray/subsasgn
Matrix index is out of range for deletion.

Error in triageTemplates2 (line 7)
W(:,idrop,:) = [];

Error in learnAndSolve8b (line 188)
                triageTemplates2(ops, iW, C2C, W, U, dWU, mu, nsp, ndrop);

Error in MEAAnalysis/getKiloSorting (line 203)
            rez = learnAndSolve8b(rez);
```
Any ideas?

Cheers,
Noah

**Comment by NoahDolev** [comment]
Might relate also  #24

**Comment by NoahDolev** [comment]
Hi,

Small update. It seems that nsp gets the dimensions units x units instead of being a vector of size units. 

The problem occurs in line 153 of leardAndSolve8b:


```{Matlab}
 nsp = nsp * p1 + (1-p1) * double(nsp0);
```

nsp is of dimensions units x 1 but nsp0 is 1 x units. So, finally you get units x units . 

Perhaps the line should be ``` .* ```?

Best,
Noah

**Comment by NoahDolev** [comment]
Nope, this wouldn't work either. nsp0 seems to flip from batch to batch. Sometimes units x 1 and sometimes 1 x units.

**Comment by marius10p** [comment]
Any idea why the aspect ratio flips? It doesn't flip for most people. I can fix it by vectorizing x = x(:), but I would like to understand why it happens (it shouldn't).

**Comment by NoahDolev** [comment]
No, but it does seem to happen to others. See the referenced issues. It's strange.

**Comment by NoahDolev** [comment]
Any clues, Marius?

Thanks again for your help!

**Comment by marius10p** [comment]
Does it get fixed by vectorizing nsp0 like this: 
nsp = nsp * p1 + (1-p1) * double(nsp0(:));

**Comment by NoahDolev** [comment]
Hi,

It gets me a few lines down to line 197:

```{Matlab}
Error using gpuArray/subsasgn
Matrix index is out of range for deletion.

Error in triageTemplates2 (line 7)
W(:,idrop,:) = [];

Error in learnAndSolve8b (line 197)
                triageTemplates2(ops, iW, C2C, W, U, dWU, mu, nsp, ndrop);
```
If I use the if statement from above,  I run into another indexing issue:

```{Matlab}
Error using gpuArray/subsref
Subscript indices must either be real positive integers or logicals.

Error in getMeWtW (line 21)
    iList = int32(gpuArray(isort(iNear, :)));

Error in learnAndSolve8b (line 178)
        [WtW, iList] = getMeWtW(single(W), single(U), Nnearest);

Error in MEAAnalysis/getKiloSorting (line 205)
            rez = learnAndSolve8b(rez);
```
In this case, it seems isort is a gpuArray of size 0 x 0 . 

What do you think?

Best,
Noah

**Comment by facazevedo** [comment]
Any news about how the fix the error? I am also having the same problem with my own linear 16 ch dataset. In order to reproduce it, I simulated 40 s of data using "make_eMouseData_drift.m" and the provided "Linear16x1_kilosortChanMap.mat".  Then, I got the following and possibly related error: 

```
>> master_kilosort
Time   0s. Determining good channels.. 
found 1760 threshold crossings in 40.00 seconds of data 
found 7 bad channels 
Time   3s. Computing whitening matrix.. 
Getting channel whitening matrix... 
Channel-whitening filters computed. 
Time   4s. Loading raw data and applying filters... 
Time   4s. Finished preprocessing 19 batches. 
Obtained 7 PC waveforms in 0.23 seconds 
time 0.30, pre clustered 1 / 19 batches 
time 0.02, compared 1 / 19 batches 
time 2.77, Re-ordered 19 batches. 
Time   3s. Optimizing templates ...
3.19 sec, 1 / 39 batches, 3 units, nspks: 1.5768, mu: 14.6883, nst0: 29, merges: 0.0000, 0.0000 
memorized middle timepoint 
reverted back to middle timepoint 
Elapsed time is 4.628088 seconds.

ntot  = 2462

Error using gpuArray/subsref
Index exceeds matrix dimension.

Error in learnAndSolve8b (line 351)
    rez.W_a(:,:,j) = gather(A(:, 1:nKeep) * B(1:nKeep, 1:nKeep));

Error in master_kilosort_temp (line 39)
rez = learnAndSolve8b(rez);
```
Related m-files can be found here: 
[error_kilosort2_fred.zip](https://github.com/MouseLand/Kilosort2/files/3081094/error_kilosort2_fred.zip)

Best.

**Comment by marius10p** [comment]
Sorry, I haven't had time to look at this yet, but soon! It sounds like some batches don't have spikes at all. 

@facazevedo I am not sure your problem is the same, but I would always test on longer recordings, say >10 minutes, since strange edge cases might occur with very short data files.

**Comment by NoahDolev** [comment]
FYI: I have now tried restarting Matlab for several different recordings and I get the same behavior. Sometimes successful completion and sometimes crashes.

**Comment by zekearneodo** [comment]
Hi, I get the same error on the GUI, and when I run it from a master file, I get the error

```
Error using gpuArray/subsasgn
Attempt to grow array along ambiguous dimension.

Error in learnAndSolve8b (line 205)
            nsp(Nfilt + [1:size(dWU0,3)]) = ops.minFR * NT/ops.fs;

Error in master (line 31)
rez = learnAndSolve8b(rez);
```
Could this be related?

**Comment by sujayane** [comment]
Any update on this error? I am also getting it (with 32chan V-probe data).

**Comment by Lidors** [comment]
I get the same error as Noah described in his first comment (using a 6-shank probe with 64 channels).
Any ideas on how to solve it?
Thanks, Lidor

**Comment by czuba** [comment]
Still running into this bug. In my hands, it pops up when ops.minFR is set too low, reinforcing notion that it's related to units/batches without any spikes. Probably somewhere in code where a idrop (or nsp, or some other predecessor) is getting initialized with improper dimensions but normally gets overwritten by properly formatted return arguments. 

Set ops.minFR = 0 and you can probably replicate error on just about any dataset.

**Comment by Alkohost** [comment]
This problem can be solved sometimes by changing the batch size or the detection threshold, but it still appears unpredictably. In some cases I add to data big number of fictive spikes, but the error is still there.

**Comment by kalfton** [comment]
I dug into this problem, and I think I got the reason.
There is a variable called "nsp" in  "learnAndSolve8b.m",  this variable should be a n*1 gpuarray. But sometime it becomes a 1*n gpuarray. I think someone has mentioned above that nsp becomes a n*n matrix. This is because the code: nsp = nsp * p1 + (1-p1) * double(nsp0); in line 153, which add a 1*n array with a n*1 array.

Why would nsp variable become 1*n? This is because of the code in line 204: 
nsp(Nfilt + [1:size(dWU0,3)]) = ops.minFR * NT/ops.fs;
This code intends to extend nsp array from n*1 to (n+m)*1, but the problem is if nsp is a 1*1 matrix, the direction it extend is in row direction, i.e. from 1*1 to 1*(m+1). 

Here is the way I solve it. Before line 153: nsp = nsp * p1 + (1-p1) * double(nsp0); I add the code:
if size(nsp,2)>1
        nsp = nsp';
end
This is to check if nsp is a column vector, if not, transpose it to be a column vector.
Or a easier way is replacing line nsp(Nfilt + [1:size(dWU0,3)]) = ops.minFR * NT/ops.fs; to be 
nsp(Nfilt + [1:size(dWU0,3)], :) = ops.minFR * NT/ops.fs; (Though I didn't test it on my own code).

@marius10p Could you confirm if what I wrote is correct?

Best,

Kaining

**Comment by marius10p** [comment]
Looks like this was added to the code, closing issue.

**Comment by johnmbarrett** [comment]
Ran into another situation where zekearneodo's error (https://github.com/MouseLand/Kilosort/issues/44#issuecomment-493228843) can occur, when nsp becomes empty (size 0x1). In that case the guard on line 190 of mainLoop/learnTemplates.m fails (all line numbers based on commit 57cd4ad as I had to roll back to that one due to issue https://github.com/MouseLand/Kilosort/issues/320). Replacing

`if nsprow<nspcol`

with

`if isrow(nsp)`

fixes it, but then you also have to guard against CUDA kernel errors that occur when assigning into or computing the median of an empty gpuArray on lines 244, 245, and 261. That gets you out of the main loop, although then I ran into further downstream errors (specifically an array dimension mismatch on line 105 of track and sort) and gave up because I think there's actually something wrong with my data files.

Anywho, point is there's a still an edge case where this issue can occur. I can send you the offending files if you want to try and debug, but they're quite large (~4 GB).
